### PR TITLE
bpo-42202: Store func annotations as single tuple at bytecode level

### PR DESCRIPTION
**Pull request by uriyyo**
Store func annotations as single tuple at bytecode level.

https://bugs.python.org/issue42202

### CODE REVIEW

File: Objects/codeobject.c
Comment (serhiy-storchaka) [inline Objects/codeobject.c]:
```suggestion
    if (annotations && annotations != Py_None && !PyTuple_Check(annotations)) {
```

File: Objects/codeobject.c
Comment (serhiy-storchaka) [inline Objects/codeobject.c]:
```suggestion
                "code: annotations must be tuple or None");
```

File: Objects/funcobject.c
Comment (serhiy-storchaka) [inline Objects/funcobject.c]: Style nit:
```suggestion
        PyObject *annotations = ((PyCodeObject*) op->func_code )->co_annotations;
```

File: Python/compile.c
Comment (serhiy-storchaka) [inline Python/compile.c]: `PyTuple_Pack()` will crash if `_PyAST_ExprAsUnicode()` returned NULL. And there is a leak.

Check the result of `_PyAST_ExprAsUnicode()`. And decref it after use.

File: Python/compile.c
Comment (serhiy-storchaka) [inline Python/compile.c]: Check the result and decref annotations.

File: Python/compile.c
Comment (serhiy-storchaka) [inline Python/compile.c]: No longer need to distinguish 1 and -1. Always return 1 on success.

File: Python/compile.c
Comment (serhiy-storchaka) [inline Python/compile.c]:
```suggestion
    if (!compiler_visit_annotations(c, args, returns)) {
```

File: PC/launcher.c
Comment (serhiy-storchaka) [inline PC/launcher.c]: This is a range. No need to change it. But you need to increase the magic number in `Lib/importlib/_bootstrap_external.py`.

File: Include/cpython/code.h
Comment (serhiy-storchaka) [inline Include/cpython/code.h]: I have doubts about changing the C API. See, we have `PyCode_New` and `PyCode_NewWithPosOnlyArgs` because of new parameter. You need to introduce a new constructor function, or call the old function and change the code object after this with a new setter function.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
LGTM in general.

File: Include/cpython/code.h
Comment (uriyyo) [reply Include/cpython/code.h]: Totally agree with you, it will be better to be backward compatible with previous versions.
I have added function `PyCode_NewWithAnnotations` in order to create `CodeObject` with annotations.

File: PC/launcher.c
Comment (uriyyo) [reply PC/launcher.c]: Fixed, my bad.

File: Python/compile.c
Comment (uriyyo) [reply Python/compile.c]: Fixed, thanks.

File: Python/compile.c
Comment (uriyyo) [reply Python/compile.c]: Fixed, thanks.

File: Python/compile.c
Comment (uriyyo) [reply Python/compile.c]: Fixed, thanks.

**Comment by uriyyo** [comment]
@serhiy-storchaka I have implemented the same logic for module and class annotations, so all annotations will be computed at compilation time.

File: Lib/test/test_opcodes.py
Comment (uriyyo) [inline Lib/test/test_opcodes.py]: This test is not valid anymore because all annotations values will be set within `SETUP_ANNOTAITONS` opcode.

**Comment by uriyyo** [comment]
I have made the requested changes; please review again

**Comment by uriyyo** [comment]
I have made the requested changes; please review again

**Comment by uriyyo** [comment]
@serhiy-storchaka I have reimplemented this feature and remove `co_annotations` field from `CodeObject`.

New implementation follows the idea provided by Inada Naoki at issue discussion https://bugs.python.org/issue42202#msg381309

It's a pretty easy feature and will speedup function declaration with annotations more than x2 times.

File: Objects/funcobject.c
Comment (methane) [inline Objects/funcobject.c]:
```suggestion
    if (PyTuple_CheckExact(op->func_annotations)) {
```

Put space after `)`.

File: Objects/funcobject.c
Comment (methane) [inline Objects/funcobject.c]:
```suggestion
        if (op->func_annotations == NULL) {
            Py_DECREF(annotations);
            return NULL;
        }
```

Use `{` always.

File: Objects/funcobject.c
Comment (methane) [reply Objects/funcobject.c]: And you can use shorter variable name for local variables.

File: Objects/funcobject.c
Comment (methane) [inline Objects/funcobject.c]:
```suggestion
        for (Py_ssize_t i = 0; i < PyTuple_Size(annotations); i += 2) {
            int err = PyDict_SetItem(op->func_annotations,
```

File: Objects/funcobject.c
Comment (methane) [inline Objects/funcobject.c]:
```suggestion
                Py_DECREF(op->func_annotations);
                Py_DECREF(annotations);
```

File: Python/compile.c
Comment (methane) [inline Python/compile.c]:
```suggestion
        if (!value) {
```

**Review by methane — CHANGES_REQUESTED** [review]

**Comment by uriyyo** [comment]
I have made the requested changes; please review again

File: Python/compile.c
Comment (methane) [inline Python/compile.c:2034]: Remove compiler_visit_annexpr.

Comment (uriyyo) [reply Python/compile.c:2034]: This function is used for class/module annotations.

Comment (uriyyo) [reply Python/compile.c:2034]: This is a line where it used
https://github.com/python/cpython/blob/ce04e7105bc396c32667a22b928a712ba0778a3f/Python/compile.c#L5271

Please correct me if I miss smth.

Comment (methane) [reply Python/compile.c:2034]: You are right. I grepped `compiler_visit_annexpr` by mistake. Please forget it.

File: Objects/funcobject.c
Comment (methane) [inline Objects/funcobject.c]:
```suggestion
                                     PyTuple_GET_ITEM(anns, i),
                                     PyTuple_GET_ITEM(anns, i + 1));
```

**Review by methane — APPROVED** [review]
LGTM. But wait for another coredev review.

File: Objects/funcobject.c
Comment (ZackerySpytz) [inline Objects/funcobject.c]: Please use the macro `PyTuple_GET_SIZE()` instead of `PyTuple_Size()`, as `anns` is guaranteed to be a tuple here.

File: Objects/funcobject.c
Comment (ZackerySpytz) [inline Objects/funcobject.c]: Ditto.

File: Objects/funcobject.c
Comment (ZackerySpytz) [inline Objects/funcobject.c]: `func_annotations` is already decrefed when `func_dealloc()` is called.

File: Objects/funcobject.c
Comment (uriyyo) [reply Objects/funcobject.c]: Thanks, didn't notice such difference between `PyTuple_GET_SIZE()` and `PyTuple_Size()`

File: Objects/funcobject.c
Comment (uriyyo) [reply Objects/funcobject.c]: Done

File: Objects/funcobject.c
Comment (uriyyo) [reply Objects/funcobject.c]: Removed, thanks

**Comment by ZackerySpytz** [comment]
I like this PR.

File: Objects/funcobject.c
Comment (methane) [inline Objects/funcobject.c]: When some error happened in this `if` block, `op->func_annotations` can be incomplete dict.
Then, when `func.__annotations__` is accessed next time, the incomplete dict can be returned.

To avoid returning incomplete dict:

```c
PyObject *anns = PyDict_New();

// fill anns...

// finally
Py_SETREF(op->func_annotations, anns);
```

Comment (uriyyo) [reply Objects/funcobject.c]: Nice catch, you are totally right. And thank you for the solution.

**Comment by methane** [comment]
Would you update documents?

* what's new: Add an entry in "Optimizations" section that describe performance/memory saving.
* what's new: Add an entry in "Porting to Python 3.10" section that describe bytecode change. You need to add "CPython bytecode changes" section. ([See Python 3.9 what's new](https://docs.python.org/dev/whatsnew/3.9.html#cpython-bytecode-changes).)
* library/dis: Update `MAKE_FUNCTION` document, and add ``versionchanged: 3.10` block.

**Comment by uriyyo** [comment]
> Would you update documents?
> 
> * what's new: Add an entry in "Optimizations" section that describe performance/memory saving.
> * what's new: Add an entry in "Porting to Python 3.10" section that describe bytecode change. You need to add "CPython bytecode changes" section. ([See Python 3.9 what's new](https://docs.python.org/dev/whatsnew/3.9.html#cpython-bytecode-changes).)
> * library/dis: Update `MAKE_FUNCTION` document, and add ``versionchanged: 3.10` block.

Sure, I will update docs regarding this PR.

**Comment by methane** [comment]
This pull request also break backward compatibility of bytecode. Bump magic number.

See https://devguide.python.org/compiler/?highlight=bytecode#introducing-new-bytecode for how you can bump the magic number.

**Comment by uriyyo** [comment]
@methane I have updated docs, can you please review it?

**Comment by markshannon** [comment]
Rather than build a list of strings, and then emit `LOAD_CONST tuple(annotation)`, the code would be cleaner if `compiler_visit_argannotation` emitted code a `LOAD_CONST annotation_string` for each annotation.
The CFG optimizer already has to code to clean this up, so it is probably best to not duplicate this effort.

No need to do anything for this PR though, a separate PR would probably be best.

File: Doc/library/dis.rst
Comment (methane) [inline Doc/library/dis.rst]:
```suggestion
   .. versionchanged:: 3.10
      Flag value ``0x04`` is a tuple of strings instead of dictionary
```

**Comment by methane** [comment]
@markshannon Thank you for your advice. It is nice idea.

@uriyyo May I merge this PR as current code? Or do you want to simplify the code as Mark Shannon suggested?

**Comment by uriyyo** [comment]
@methane I will try to simplify code to not create separate PR.

**Comment by uriyyo** [comment]
@methane @markshannon 
I have updated PR regarding Mark advice, can you please review PR again?

**Review by methane — APPROVED** [review]

**Comment by uriyyo** [comment]
@methane Should Mark review this PR or we can merge it?

**Comment by methane** [comment]
@markshannon Would you review this? If no, I will merge this after another look in this week.

File: Objects/funcobject.c
Comment (methane) [inline Objects/funcobject.c]: This empty line looks bit ugly.

File: Objects/funcobject.c
Comment (methane) [inline Objects/funcobject.c]: Use `{}` always in new code.

File: Python/compile.c
Comment (methane) [inline Python/compile.c]: Can we just `return 0` instead of `goto error;`?

**Review by methane — CHANGES_REQUESTED** [review]

File: Objects/funcobject.c
Comment (uriyyo) [reply Objects/funcobject.c]: Agree, fixed

File: Objects/funcobject.c
Comment (uriyyo) [reply Objects/funcobject.c]: My bad, I didn't know about that

File: Python/compile.c
Comment (uriyyo) [reply Python/compile.c]: Sure, fixed

**Comment by uriyyo** [comment]
I have made the requested changes; please review again

**Comment by markshannon** [comment]
Rather than return the tuple size via a pointer (out) parameter, you can return the size directly as it cannot be negative, leaving -1 to signify an error. So, 
```C
int compiler_visit_argannotation(struct compiler *c, identifier id, expr_ty annotation, Py_ssize_t *annotations_len);
```
becomes
```C
int compiler_visit_argannotation(struct compiler *c, identifier id, expr_ty annotation);
```
and the use changes from
```C
if (compiler_visit_argannotation(..., &len))
    goto error;
```
to 
```C
len = compiler_visit_argannotation(...);
if (len < 0)
    goto error;
```

**Comment by uriyyo** [comment]
> Rather than return the tuple size via a pointer (out) parameter, you can return the size directly as it cannot be negative, 
> leaving -1 to signify an error.

I think it will make code more complicated because it will require rewriting all ifs for `compiler_visit_annotations` function.

It will be great if `compiler_visit_argannotations` and `compiler_visit_argannotation` will follow the same conventions - return tuple size through the pointer or as return value.

@markshannon  What is your opinion regarding that?

**Comment by markshannon** [comment]
I prefer my approach (I wouldn't have suggested it otherwise :slightly_smiling_face: ), but the code is fine as it is.

I'm still happy with merging it.

**Comment by uriyyo** [comment]
I totally agree with you, personally, I don't like when the return value passed through the pointer instead of simple return, as zen said `explicit is better than implicit`.

But otherwise, In my opinion, it will make code less readable.
From this:
```c
if (!compiler_visit_argannotations(c, args->args, &annotations_len))
   return 0;
if (args->vararg && args->vararg->annotation &&
    !compiler_visit_argannotation(c, args->vararg->arg,
                                  args->vararg->annotation, &annotations_len))
   return 0;
```
It will be rewritten to:
```c
    len = compiler_visit_argannotations(c, args->args);
    if (len < 0)
        return 0;
    annotations_len += len;
    if (args->vararg && args->vararg->annotation) {
        len = compiler_visit_argannotation(c, args->vararg->arg,
                                           args->vararg->annotation);
        if (len < 0)
            return 0;
        annotations_len += len;
    }
```

This situation is a double-edged sword when we should choose between code readability and code explicity.

File: Doc/whatsnew/3.10.rst
Comment (Fidget-Spinner) [inline Doc/whatsnew/3.10.rst]:
```suggestion
* Function parameters and their annotations are no longer computed at runtime, 
  but rather at compilation time.  They are stored as a tuple of strings at the 
  bytecode level.  It is now around 100% faster to create a function with parameter 
  annotations.  (Contributed by Yurii Karabas and Inada Naoki in :issue:`42202`)
```

**Review by Fidget-Spinner — COMMENTED** [review]
I don't have anything valuable to add here, just a small nitpick with the whatsnew.

This optimization makes me really happy :). Thanks Yuri for the PR and Inada-san for the concept.

File: Doc/whatsnew/3.10.rst
Comment (uriyyo) [reply Doc/whatsnew/3.10.rst]: Thanks💫

**Comment by uriyyo** [comment]
@Fidget-Spinner Thank you for making docs more clear and for adding Inada to the list of contributors. It's my bad, I must definitely add Inada to the contributor's list because it's fully his idea.

**Comment by uriyyo** [comment]
Can someone rerun Azure Pipelines?
I believe it failed not because of my code)

**Comment by methane** [comment]
Even core committers can not re-run Azure Pipeline. Close and reopen is the most easy way to rebuild, although it starts not only Azure Pipeline.

**Comment by uriyyo** [comment]
@methane Looks like everything is done and this PR can be merged

**Comment by methane** [comment]
I did `make regen-importlib` but I could not `git push` to your repository. Would you do it?

**Comment by uriyyo** [comment]
> I did make regen-importlib but I could not git push to your repository. Would you do it?

Sure, no problems.

**Comment by uriyyo** [comment]
@methane I have regenerated `import-lib`

**Comment by methane** [comment]
Congrats, @uriyyo. And thank you reviewers.

**Comment by uriyyo** [comment]
Thanks all, I am happy that we merge this PR💫

@methane Should we do smth similar for class/module annotations?

**Pull request merged**
